File: Examples/Python/Containers/MongoDB-REST/src/controllers/health_controller.py
import time
import os
from flask import Blueprint, jsonify, current_app

from src.services.MongoDB.REST.mongodb_rest_client import MongoDB_REST_Client
from src.health.health_result_entry import HealthResultEntry
from src.health.health_results import HealthResults

health_blueprint = Blueprint('health', __name__)

def build_health_check_mongodb(mongodb_rest_client: MongoDB_REST_Client) -> HealthResultEntry:
    start = time.time()
    try:
        mongodb_rest_client.ping()
        return HealthResultEntry('mongodb', True, round((time.time()-start) * 1000))
    except Exception as e:
        return HealthResultEntry('mongodb', False, round((time.time()-start) * 1000), str(e))
    

def build_health_check_response(mongodb_rest_client: MongoDB_REST_Client) -> HealthResults:
    start = time.time()
    results = [build_health_check_mongodb(mongodb_rest_client)]
    return HealthResults(round((time.time()-start) * 1000), results)

@health_blueprint.route('/', methods=['GET', 'POST'])
def health_check():
    try:
        return jsonify(build_health_check_response(current_app.mongodb_rest_client).to_dict())
    except Exception as e:
        return str(e), 500

File: Examples/Rust/Containers/MongoDB-REST/src/src/controllers/health_controller.rs
use actix_web::{web, HttpResponse, Responder};
use std::time::Instant;
use tokio::join;

use crate::services::mongodb::rest::mongodb_rest_client::MongoDBRESTClient;
use crate::health::health_results::HealthResults;
use crate::health::health_result_entry::HealthResultEntry;

/// Asynchronously checks MongoDB health.
async fn build_health_check_mongodb(mongodb_rest_client: web::Data<MongoDBRESTClient>) -> HealthResultEntry {
    let start = Instant::now();
    match mongodb_rest_client.ping().await {
        Ok(..) => HealthResultEntry::new("mongodb".to_string(), true, start.elapsed().as_millis() as u64, None),
        Err(e) => HealthResultEntry::new("mongodb".to_string(), false, start.elapsed().as_millis() as u64, Some(e.to_string())),
    }
}

/// Builds the full health check response, running checks in parallel.
async fn build_health_check_response(mongodb_rest_client: web::Data<MongoDBRESTClient>) -> HealthResults {
    let start = Instant::now();
    // Run checks concurrently using `join!`
    let (mongodb_result,) = join!(build_health_check_mongodb(mongodb_rest_client));
    HealthResults::new(start.elapsed().as_millis() as u64, vec![mongodb_result])
}

/// Health check endpoint
async fn health_check(mongodb_rest_client: web::Data<MongoDBRESTClient>) -> impl Responder {
    let health_results = build_health_check_response(mongodb_rest_client).await;
    if health_results.healthy {
        HttpResponse::Ok().json(health_results)
    } else {
        HttpResponse::InternalServerError().json(health_results)
    }
}

/// Registers routes in Actix-web
pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::scope("/api/health")
            .route("", web::get().to(health_check))
            .route("", web::post().to(health_check))
        );
}
